src/get_imports/mod.rs

```diff
@@ -73,78 +70,61 @@ fn get_imports_from_file<'a>(
+
+    let globals = Globals::new();
+    GLOBALS.set(&globals, || {
+        let mut resolver = swc_core::transforms::resolver(
+            Mark::fresh(Mark::root()),
+            Mark::fresh(Mark::root()),
+            true,
+        );
+        let resolved = fold_module(&mut resolver, ts_module.clone());
+        visit_module(&mut visitor, &resolved);
+    });
+    let imports_map = get_imports_map_from_visitor(visitor);
 
     return Ok(imports_map);
 }
 
-fn capture_imports_map(
-    ts_module: swc_ecma_ast::Module,
-    fm: Arc<SourceFile>,
+fn get_imports_map_from_visitor(
+    visitor: ImportPathVisitor,
 ) -> HashMap<String, Option<HashSet<String>>> {
-    let mut imports_map: HashMap<String, Option<HashSet<String>>> = HashMap::new();
-    ts_module.body.iter().for_each(|node| {
-        if node.is_module_decl() {
-            if let Some(module_decl) = node.as_module_decl() {
-                if module_decl.is_import() {
-                    // let source_text = &fm.src.as_bytes().to_vec();
-                    let i = module_decl.as_import().unwrap(); // Safe to unwrap due to previous is_import assertion
-                    let specs_name: Vec<String> = i
-                        .specifiers
-                        .iter()
-                        .filter_map(|spec| -> Option<String> {
-                            return get_specifier_name(&fm, spec);
-                        })
-                        .collect();
-                    let import_source = i.src.value.to_string();
-                    match imports_map.get(&import_source) {
-                        Some(Some(current_set)) => {
-                            // TODO find a way to append data to current set instead of copying it into an new one
-                            let mut new_set: HashSet<String> =
-                                HashSet::from_iter(current_set.iter().map(|v| v).cloned());
-                            for val in specs_name {
-                                new_set.insert(val);
-                            }
-                            imports_map.insert(import_source, Some(new_set));
-                        }
-                        _ => {
-                            if specs_name.is_empty() {
-                                imports_map.insert(import_source, None);
-                            } else {
-                                let new_set: HashSet<String> =
-                                    HashSet::from_iter(specs_name.iter().cloned());
-                                imports_map.insert(import_source, Some(new_set));
-                            }
-                        }
-                    }
+    let mut final_imports_map: HashMap<String, Option<HashSet<String>>> = HashMap::new();
+    visitor
+        .imports_map
+        .iter()
+        .for_each(|(k, v)| match final_imports_map.get_mut(k) {
+            Some(Some(specifiers)) => {
+                for spec in v {
+                    specifiers.insert(spec.clone());
```

Comment:
I think you can avoid these clones.

- first, move the struct fields onto the stack individually by destructuring:

https://doc.rust-lang.org/rust-by-example/scope/move/partial_move.html

- then, instead of iterating over the data structures, drain them

https://doc.rust-lang.org/stable/std/vec/struct.Drain.html

https://doc.rust-lang.org/stable/std/collections/hash_map/struct.Drain.html

https://doc.rust-lang.org/stable/std/collections/hash_set/struct.Drain.html

Reply:
(same for import_paths and require_paths below)